node/router/src/handshake.rs

```diff
@@ -36,44 +36,92 @@ use tokio::net::TcpStream;
 use tokio_stream::StreamExt;
 use tokio_util::codec::Framed;
 
+/// A trait that enables wrapping custom handshake logic within the router logic.
+///
+/// This keeps peer collections nicely encapsulated with nicer error handling.
+#[async_trait]
+pub trait ExtendedHandshake<N: Network>: Handshake + Outbound<N> {
+    /* User implemented methods. */
+
+    fn genesis_header(&self) -> io::Result<Header<N>>;
+
+    async fn custom_handshake<'a>(
+        &'a self,
+        _conn_addr: SocketAddr,
+        peer: Peer<N>,
+        framed: Framed<&'a mut TcpStream, MessageCodec<N>>,
+    ) -> io::Result<(Peer<N>, Framed<&'a mut TcpStream, MessageCodec<N>>)> {
+        Ok((peer, framed))
```

Comment:
we could use a comment here that by default there is no customization, hence this default method

Reply:
also, does this one need to return the `peer` and `framed`? those should already be available after the base handshake